File: scm-csharp/scheme/TextOutputStream.cs
namespace scheme;

public class TextOutputStream : TextWriter
{
    private TextWriter writer;
    public bool IsOpen { get; private set; } = true;

    public TextOutputStream(TextWriter writer)
    {
        this.writer = writer;
    }

    public TextWriter Inner => writer;

    public override System.Text.Encoding Encoding => writer.Encoding;

    public override void Write(char value)
    {
        writer.Write(value);
    }

    public override void Write(char[] buffer, int index, int count)
    {
        writer.Write(buffer, index, count);
    }

    public override void Write(string? value)
    {
        writer.Write(value);
    }

    public override void Flush()
    {
        writer.Flush();
    }

    public override void Close()
    {
        IsOpen = false;
        writer.Close();
    }

    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            IsOpen = false;
            writer.Dispose();
        }
        base.Dispose(disposing);
    }

    public override string ToString()
    {
        return writer.ToString()!;
    }
}

File: scm-java/src/main/scheme/TextOutputStream.java
package scheme;

import java.io.IOException;
import java.io.Writer;

public class TextOutputStream extends Writer {
    private Writer writer;
    public boolean isOpen = true;

    public TextOutputStream(Writer writer) {
        this.writer = writer;
    }

    public Writer getInner() {
        return writer;
    }

    private void ensureOpen() throws IOException {
        if (!isOpen) throw new IOException("port is closed");
    }

    @Override
    public void write(char[] cbuf, int off, int len) throws IOException {
        ensureOpen();
        writer.write(cbuf, off, len);
    }

    @Override
    public void write(int c) throws IOException {
        ensureOpen();
        writer.write(c);
    }

    @Override
    public void write(String str, int off, int len) throws IOException {
        ensureOpen();
        writer.write(str, off, len);
    }

    @Override
    public void flush() throws IOException {
        writer.flush();
    }

    @Override
    public void close() throws IOException {
        isOpen = false;
        writer.close();
    }

    @Override
    public String toString() {
        return writer.toString();
    }
}
